pkg/actors/actors.go

```diff
@@ -582,64 +602,145 @@ func (a *actorsRuntime) callLocalActor(ctx context.Context, req *invokev1.Invoke
-	if appCh == nil {
+	if a.appChannel == nil {
 		return nil, fmt.Errorf("app channel for actor type %s is nil", act.actorType)
 	}
 
 	policyDef := a.resiliency.ActorPostLockPolicy(act.actorType, act.actorID)
 
 	// If the request can be retried, we need to enable replaying
 	if policyDef != nil && policyDef.HasRetries() {
-		req.WithReplay(true)
+		imReq.WithReplay(true)
 	}
 
 	policyRunner := resiliency.NewRunnerWithOptions(ctx, policyDef,
 		resiliency.RunnerOpts[*invokev1.InvokeMethodResponse]{
 			Disposer: resiliency.DisposerCloser[*invokev1.InvokeMethodResponse],
 		},
 	)
-	resp, err := policyRunner(func(ctx context.Context) (*invokev1.InvokeMethodResponse, error) {
-		return appCh.InvokeMethod(ctx, req, "")
+	imRes, err := policyRunner(func(ctx context.Context) (*invokev1.InvokeMethodResponse, error) {
+		return a.appChannel.InvokeMethod(ctx, imReq, "")
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	if resp == nil {
+	if imRes == nil {
 		return nil, errors.New("error from actor service: response object is nil")
 	}
+	defer imRes.Close()
 
-	if resp.Status().GetCode() != http.StatusOK {
-		respData, _ := resp.RawDataFull()
+	if imRes.Status().GetCode() != http.StatusOK {
+		respData, _ := imRes.RawDataFull()
 		return nil, fmt.Errorf("error from actor service: %s", string(respData))
 	}
 
-	// The .NET SDK signifies Actor failure via a header instead of a bad response.
-	if _, ok := resp.Headers()["X-Daprerrorresponseheader"]; ok {
-		return resp, actorerrors.NewActorError(resp)
+	// Get the protobuf
+	res, err := imRes.ProtoWithData()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response data: %w", err)
+	}
+
+	// The .NET SDK indicates Actor failure via a header instead of a bad response
+	if _, ok := res.GetHeaders()["X-Daprerrorresponseheader"]; ok {
+		return res, actorerrors.NewActorError(res)
+	}
+
+	return res, nil
+}
+
+// Locks an internal actor for a request
+//
+//nolint:protogetter
+func (a *actorsRuntime) lockInternalActorForRequest(act *actor, req *internalv1pb.InternalInvokeRequest) (md map[string][]string, err error) {
+	var reentrancyID *string
+	// The req object is nil if this is a request for a reminder or timer
+	if req != nil {
+		// Get metadata in a map
+		// Allocate with an extra 1 capacity to add the Dapr-Reentrancy-Id value if needed
+		md = make(map[string][]string, len(req.Metadata)+1)
+		for k, v := range req.Metadata {
+			vals := v.GetValues()
+			if len(vals) == 0 {
+				continue
+			}
+			md[k] = vals
+		}
+	}
+
+	// Reentrancy to determine how we lock.
+	if a.actorsConfig.GetReentrancyForType(act.actorType).Enabled {
+		if md != nil && len(md["Dapr-Reentrancy-Id"]) > 0 {
+			reentrancyID = ptr.Of(md["Dapr-Reentrancy-Id"][0])
+		} else {
+			var uuidObj uuid.UUID
+			uuidObj, err = uuid.NewRandom()
+			if err != nil {
+				return nil, fmt.Errorf("failed to generate UUID: %w", err)
+			}
+			reentrancyID = ptr.Of(uuidObj.String())
+			if md == nil {
```

Comment:
md is already initialized with length, 1 greater than metadata length - do we need this check here and re-initialize it?

Reply:
We can't be sure md != nil here